# **1 – Introduction**

# Missing Value Handling - Beginner Guide

In this notebook you will learn:

✅ What are Missing Values

✅ Why Missing Values are a Problem

✅ How to Detect Missing Values

✅ How to Create Missing Values

✅ How to Handle Missing Values

✅ Mean Imputation

✅ Median Imputation

✅ Mode Imputation

Target Audience:

- School Students
- Python Beginners
- Data Analyst Aspirants

# **2 – Real Life Example**

# What are Missing Values?

Imagine a school report card.

| Student | Maths |
|----------|--------|
| Ram | 90 |
| John | 80 |
| Ravi | ? |

Ravi's mark is missing.

This missing value is represented as:

NaN

NaN = Not a Number

# **3 – Load Dataset**

In [4]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [5]:
import pandas as pd

car_data = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Datasets/auto_mpg_binarised.csv')

car_data.head()

,cylinders,displacement,horsepower,weight,acceleration,model_year,origin,car_name,mpg_high
0,8,307.0,130.0,3504.0,12.0,70,1,chevrolet chevelle malibu,0
1,8,350.0,165.0,3693.0,11.5,70,1,buick skylark 320,0
2,8,318.0,150.0,3436.0,11.0,70,1,plymouth satellite,0
3,8,304.0,150.0,3433.0,12.0,70,1,amc rebel sst,0
4,8,302.0,140.0,3449.0,10.5,70,1,ford torino,0


# **4 – Check Missing Values**

In [6]:
car_data.isnull().sum()

,0
cylinders,0
displacement,0
horsepower,0
weight,0
acceleration,0
model_year,0
origin,0
car_name,0
mpg_high,0


# What happened?

isnull()

Checks whether a value is missing.

sum()

Counts how many missing values exist in each column.

# **5 – Total Missing Values**

In [7]:
total_missing_values = car_data.isnull().sum().sum()

print(total_missing_values)

0


# Breakdown of code

car_data.isnull().**sum()**

returns missing count of each column

car_data.isnull().sum().**sum()**

adding all the missing count to find the total missing value

# **7 – Create Dataset Copy**

In [8]:
car_data_copy = car_data.copy()
car_data_copy

,cylinders,displacement,horsepower,weight,acceleration,model_year,origin,car_name,mpg_high
0,8,307.0,130.0,3504.0,12.0,70,1,chevrolet chevelle malibu,0
1,8,350.0,165.0,3693.0,11.5,70,1,buick skylark 320,0
2,8,318.0,150.0,3436.0,11.0,70,1,plymouth satellite,0
3,8,304.0,150.0,3433.0,12.0,70,1,amc rebel sst,0
4,8,302.0,140.0,3449.0,10.5,70,1,ford torino,0
...,...,...,...,...,...,...,...,...,...
393,4,140.0,86.0,2790.0,15.6,82,1,ford mustang gl,1
394,4,97.0,52.0,2130.0,24.6,82,2,vw pickup,1
395,4,135.0,84.0,2295.0,11.6,82,1,dodge rampage,1
396,4,120.0,79.0,2625.0,18.6,82,1,ford ranger,1


Never modify the original dataset.

Always work on a copy.

# **8 – Import NumPy**

In [9]:
import numpy as np

# **9 – Select Numeric Columns**

In [10]:
numeric_columns = car_data_copy.select_dtypes(include=np.number).columns


numeric_columns = ['displacement', 'horsepower', 'weight', 'acceleration']
numeric_columns

['displacement', 'horsepower', 'weight', 'acceleration']

Only numerical columns are selected.

Example:

* weight
* horsepower
* acceleration

# **10 – Create 10% Missing Values**

In [11]:
np.random.seed(45)
ten_percent_car_data_missing = int(0.1 * len(car_data_copy))

for column in numeric_columns:
  random_rows = np.random.choice(car_data_copy.index, ten_percent_car_data_missing, replace=False)
  car_data_copy.loc[random_rows, column] = np.nan

car_data_copy


,cylinders,displacement,horsepower,weight,acceleration,model_year,origin,car_name,mpg_high
0,8,307.0,130.0,3504.0,12.0,70,1,chevrolet chevelle malibu,0
1,8,350.0,NaN,3693.0,11.5,70,1,buick skylark 320,0
2,8,318.0,150.0,3436.0,11.0,70,1,plymouth satellite,0
3,8,304.0,NaN,3433.0,12.0,70,1,amc rebel sst,0
4,8,NaN,140.0,3449.0,10.5,70,1,ford torino,0
...,...,...,...,...,...,...,...,...,...
393,4,140.0,86.0,2790.0,NaN,82,1,ford mustang gl,1
394,4,97.0,52.0,2130.0,24.6,82,2,vw pickup,1
395,4,135.0,84.0,2295.0,11.6,82,1,dodge rampage,1
396,4,120.0,79.0,2625.0,18.6,82,1,ford ranger,1


**Step 1:**

Set a random seed value (45) so that the same rows are selected every time the code runs.

**Step 2:**

Calculate 10% of the total rows in the dataset.

For example:
If the dataset contains 398 rows,

10% of 398 = 39 rows

**Step 3:**

Loop through each numeric column in the dataset.

Examples:
- displacement
- horsepower
- weight
- acceleration

**Step 4:**

Randomly select 39 row numbers from the dataset without repeating any row.

**Step 5:**

Replace the values in those selected rows with NaN (Missing Value).

**Step 6:**

Repeat the same process for every numeric column.

**Result:**

Each numeric column will contain approximately 10% missing values.

# **11 – Verify Missing Values**

In [12]:
car_data_copy.isnull().sum()

,0
cylinders,0
displacement,39
horsepower,39
weight,39
acceleration,39
model_year,0
origin,0
car_name,0
mpg_high,0


# **12 – Visual Explanation**

Original Dataset

100%
Data Available

↓

Randomly Remove 10%

↓

Dataset with Missing Values

↓

Apply Imputation Techniques

# **13 – Mean Concept**

# Mean Imputation

Example:

10
20
30
NaN

Mean = (10+20+30)/3

Mean = 20

Replace NaN with 20

In [13]:
from sklearn.impute import SimpleImputer

mean_imputer = SimpleImputer(strategy='mean')

car_data_copy_mean = car_data_copy.copy()

car_data_copy_mean[numeric_columns] = mean_imputer.fit_transform(car_data_copy_mean[numeric_columns])

car_data_copy_mean


,cylinders,displacement,horsepower,weight,acceleration,model_year,origin,car_name,mpg_high
0,8,307.000000,130.00000,3504.0,12.000000,70,1,chevrolet chevelle malibu,0
1,8,350.000000,103.34637,3693.0,11.500000,70,1,buick skylark 320,0
2,8,318.000000,150.00000,3436.0,11.000000,70,1,plymouth satellite,0
3,8,304.000000,103.34637,3433.0,12.000000,70,1,amc rebel sst,0
4,8,192.249304,140.00000,3449.0,10.500000,70,1,ford torino,0
...,...,...,...,...,...,...,...,...,...
393,4,140.000000,86.00000,2790.0,15.582173,82,1,ford mustang gl,1
394,4,97.000000,52.00000,2130.0,24.600000,82,2,vw pickup,1
395,4,135.000000,84.00000,2295.0,11.600000,82,1,dodge rampage,1
396,4,120.000000,79.00000,2625.0,18.600000,82,1,ford ranger,1


### **13a - Mean Imputation**

This code replaces missing values (NaN) in numeric columns with the average (mean) value of that column.

**Step 1:** Import `SimpleImputer`, a tool used to fill missing values.

**Step 2:** Create a Mean Imputer by setting `strategy='mean'`.

**Step 3:** Create a copy of the dataset to avoid modifying the original data.

**Step 4:** Calculate the mean value for each numeric column.

**Step 5:** Replace all missing values in those columns with their respective mean values.

**Step 6:** Display the updated dataset.

**Result:** All missing values in the numeric columns are replaced with the column average.


### **13b – Check Missing Values Again**

In [14]:
car_data_copy_mean.isnull().sum()

,0
cylinders,0
displacement,0
horsepower,0
weight,0
acceleration,0
model_year,0
origin,0
car_name,0
mpg_high,0


**Result:**

All missing values should become 0.

# **14 – Median Concept**

# Median Imputation

Values:

10
20
30
500

Median = 25

Median is less affected by outliers.

In [15]:
median_imputer = SimpleImputer(strategy='median')

car_data_copy_median = car_data_copy.copy()

car_data_copy_median[numeric_columns] = median_imputer.fit_transform(car_data_copy_median[numeric_columns])

car_data_copy_median

,cylinders,displacement,horsepower,weight,acceleration,model_year,origin,car_name,mpg_high
0,8,307.0,130.0,3504.0,12.0,70,1,chevrolet chevelle malibu,0
1,8,350.0,93.0,3693.0,11.5,70,1,buick skylark 320,0
2,8,318.0,150.0,3436.0,11.0,70,1,plymouth satellite,0
3,8,304.0,93.0,3433.0,12.0,70,1,amc rebel sst,0
4,8,151.0,140.0,3449.0,10.5,70,1,ford torino,0
...,...,...,...,...,...,...,...,...,...
393,4,140.0,86.0,2790.0,15.5,82,1,ford mustang gl,1
394,4,97.0,52.0,2130.0,24.6,82,2,vw pickup,1
395,4,135.0,84.0,2295.0,11.6,82,1,dodge rampage,1
396,4,120.0,79.0,2625.0,18.6,82,1,ford ranger,1


### **14a - Median Imputation**

This code replaces missing values (NaN) in numeric columns with the center (median) value of that column after it is sorted.

**Step 1:** Import `SimpleImputer`, a tool used to fill missing values.

**Step 2:** Create a Median Imputer by setting `strategy='median'`.

**Step 3:** Create a copy of the dataset to avoid modifying the original data.

**Step 4:** Calculate the median value for each numeric column.

**Step 5:** Replace all missing values in those columns with their respective median values.

**Step 6:** Display the updated dataset.

**Result:** All missing values in the numeric columns are replaced with the column center median value.

### **14b – Check Missing Values Again**

In [16]:
car_data_copy_median.isnull().sum()

,0
cylinders,0
displacement,0
horsepower,0
weight,0
acceleration,0
model_year,0
origin,0
car_name,0
mpg_high,0


**Result:**

All missing values should become 0.

# **15 – Mode Concept**

# Mode Imputation

Values:

4
4
4
6
8

Most Frequent Value = 4

Mode = 4

In [17]:
mode_imputer = SimpleImputer(strategy='most_frequent')
car_data_copy_mode = car_data_copy.copy()
car_data_copy_mode[numeric_columns] = mode_imputer.fit_transform(car_data_copy_mode[numeric_columns])
car_data_copy_mode

,cylinders,displacement,horsepower,weight,acceleration,model_year,origin,car_name,mpg_high
0,8,307.0,130.0,3504.0,12.0,70,1,chevrolet chevelle malibu,0
1,8,350.0,150.0,3693.0,11.5,70,1,buick skylark 320,0
2,8,318.0,150.0,3436.0,11.0,70,1,plymouth satellite,0
3,8,304.0,150.0,3433.0,12.0,70,1,amc rebel sst,0
4,8,97.0,140.0,3449.0,10.5,70,1,ford torino,0
...,...,...,...,...,...,...,...,...,...
393,4,140.0,86.0,2790.0,14.5,82,1,ford mustang gl,1
394,4,97.0,52.0,2130.0,24.6,82,2,vw pickup,1
395,4,135.0,84.0,2295.0,11.6,82,1,dodge rampage,1
396,4,120.0,79.0,2625.0,18.6,82,1,ford ranger,1


### **15a - Mode Imputation**

This code replaces missing values (NaN) in numeric columns with the most frequent (mode) value of that column after it is sorted.

**Step 1:** Import `SimpleImputer`, a tool used to fill missing values.

**Step 2:** Create a Median Imputer by setting `strategy='most_frequent'`.

**Step 3:** Create a copy of the dataset to avoid modifying the original data.

**Step 4:** Calculate the mode value for each numeric column.

**Step 5:** Replace all missing values in those columns with their respective mode values.

**Step 6:** Display the updated dataset.

**Result:** All missing values in the numeric columns are replaced with the column's most frequent values.

### **15b – Check Missing Values Again**

In [18]:
car_data_copy_mode.isnull().sum()

,0
cylinders,0
displacement,0
horsepower,0
weight,0
acceleration,0
model_year,0
origin,0
car_name,0
mpg_high,0


**Result:**

All missing values should become 0.

# **16 - Imputation by mode within each class**



In [19]:
car_data_copy_mode_class = car_data_copy.copy()
for car_class in car_data_copy_mode_class['mpg_high'].unique():
  class_mask = car_data_copy_mode_class['mpg_high'] == car_class
  class_df = car_data_copy_mode_class.loc[class_mask, numeric_columns]
  class_mode = class_df.mode().iloc[0]
  car_data_copy_mode_class.loc[class_mask, numeric_columns] = class_df.fillna(class_mode)
car_data_copy_mode_class

,cylinders,displacement,horsepower,weight,acceleration,model_year,origin,car_name,mpg_high
0,8,307.0,130.0,3504.0,12.0,70,1,chevrolet chevelle malibu,0
1,8,350.0,150.0,3693.0,11.5,70,1,buick skylark 320,0
2,8,318.0,150.0,3436.0,11.0,70,1,plymouth satellite,0
3,8,304.0,150.0,3433.0,12.0,70,1,amc rebel sst,0
4,8,318.0,140.0,3449.0,10.5,70,1,ford torino,0
...,...,...,...,...,...,...,...,...,...
393,4,140.0,86.0,2790.0,14.5,82,1,ford mustang gl,1
394,4,97.0,52.0,2130.0,24.6,82,2,vw pickup,1
395,4,135.0,84.0,2295.0,11.6,82,1,dodge rampage,1
396,4,120.0,79.0,2625.0,18.6,82,1,ford ranger,1


In [20]:
# 1. Create the copy
car_data_copy_mode_class = car_data_copy.copy()

# 2. Group by 'mpg_high', find the mode for each group, and fill NaNs in one shot
car_data_copy_mode_class[numeric_columns] = (
    car_data_copy_mode_class
    .groupby('mpg_high')[numeric_columns]
    .transform(lambda x: x.fillna(x.mode().iloc[0] if not x.mode().empty else np.nan))
)

car_data_copy_mode_class

,cylinders,displacement,horsepower,weight,acceleration,model_year,origin,car_name,mpg_high
0,8,307.0,130.0,3504.0,12.0,70,1,chevrolet chevelle malibu,0
1,8,350.0,150.0,3693.0,11.5,70,1,buick skylark 320,0
2,8,318.0,150.0,3436.0,11.0,70,1,plymouth satellite,0
3,8,304.0,150.0,3433.0,12.0,70,1,amc rebel sst,0
4,8,318.0,140.0,3449.0,10.5,70,1,ford torino,0
...,...,...,...,...,...,...,...,...,...
393,4,140.0,86.0,2790.0,14.5,82,1,ford mustang gl,1
394,4,97.0,52.0,2130.0,24.6,82,2,vw pickup,1
395,4,135.0,84.0,2295.0,11.6,82,1,dodge rampage,1
396,4,120.0,79.0,2625.0,18.6,82,1,ford ranger,1


# **What happens here?**

## ***A. Group the data by mpg_high***

`.groupby('mpg_high')`

Suppose mpg_high has two classes:

|mpg_high |	horsepower|
----------|-------------
|High |	100|
|High |	NaN|
|High	| 120|
|Low |	80|
|Low |	NaN|
|Low |	80|

The data is split into:

- High group
- Low group

## ***B. Find the Mode of Each Numeric Column in Each Group***
`x.mode()`

For the "Low" group:

|horsepower|
|-----------|
|80 |
|NaN |
|80 |

Mode = 80 (most frequent value)

## ***C. Fill Missing Values with the Group's Mode***

`x.fillna(x.mode().iloc[0])`

Before:

|mpg_high |	horsepower|
----------|-------------
|Low |	80|
|Low |	NaN|
|Low |	80|

After:

|mpg_high |	horsepower|
----------|-------------
|Low |	80|
|Low |	80|
|Low |	80|

## ***D. Safety Check***

`if not x.mode().empty else np.nan`

If a column contains only missing values and has no mode, keep it as NaN.


# **17 – Compare Results**

In [21]:
print("Original Dataset")
print(car_data_copy.isnull().sum().sum())

print("Mean Imputation")
print(car_data_copy_mean.isnull().sum().sum())

print("Median Imputation")
print(car_data_copy_median.isnull().sum().sum())

print("Mode Imputation")
print(car_data_copy_mode.isnull().sum().sum())

print("Mode by Class Imputation")
print(car_data_copy_mode_class.isnull().sum().sum())

Original Dataset
156
Mean Imputation
0
Median Imputation
0
Mode Imputation
0
Mode by Class Imputation
0


# **18 – Exercises**

# Practice Exercises

Exercise 1

Create 20% missing values instead of 10%.

--------------------------------

Exercise 2

Find which column contains the most missing values.

--------------------------------

Exercise 3

Calculate total missing values.

--------------------------------

Exercise 4

Apply Mean Imputation.

--------------------------------

Exercise 5

Apply Median Imputation.

--------------------------------

Exercise 6

Compare Mean and Median results.

# **19 - Interview Questions**


1. What is a Missing Value?

2. What is NaN?

3. Why are Missing Values dangerous?

4. Difference between Mean and Median?

5. When should Median be preferred?

6. What is Mode?

7. What is Imputation?

8. Why do we create a copy of the dataset?

9. What is SimpleImputer?

10. Which imputation method would you choose for skewed data?

# **20 – Summary**

# What You Learned

✅ Missing Values

✅ NaN

✅ isnull()

✅ Creating Missing Values

✅ Mean Imputation

✅ Median Imputation

✅ Mode Imputation

Next Notebook:

03_Feature_Scaling_Beginner_Guide.ipynb